commit e8e47abef6cd7f9de46ce603c675b6bc478eb0e5
Author: Nitzan Mordechai <nmordech@ibm.com>
Date:   Sun May 17 06:31:47 2026 +0000

    mgr/DaemonServer: add heap command to mgr
    
    Ceph-mgr didn't had heap profile collection, the following
    commit will add admin socket command to support heap commands:
     dump|start_profiler|stop_profiler|release|get_release_rate|set_release_rate|stats
    just like OSD, MON and MDS
    
    Fixes: https://tracker.ceph.com/issues/76639
    Signed-off-by: Nitzan Mordechai <nmordech@redhat.com>

diff --git a/src/mgr/DaemonServer.cc b/src/mgr/DaemonServer.cc
index afe0a0c37fa..97a9afdfcff 100644
--- a/src/mgr/DaemonServer.cc
+++ b/src/mgr/DaemonServer.cc
@@ -49,6 +49,7 @@
 #include "common/pick_address.h"
 #include "common/TextTable.h"
 #include "crush/CrushWrapper.h"
+#include "perfglue/heap_profiler.h"
 
 #include <boost/algorithm/string.hpp>
 
@@ -269,6 +270,14 @@ int DaemonServer::init(uint64_t gid, entity_addrvec_t client_addrs)
              asok_hook,
              "show slowest recent ops, sorted by duration");
   ceph_assert(r == 0);
+  r = admin_socket->register_command(
+    "heap " \
+    "name=heapcmd,type=CephChoices,strings=" \
+    "dump|start_profiler|stop_profiler|release|get_release_rate|set_release_rate|stats " \
+    "name=value,type=CephString,req=false",
+    asok_hook,
+    "show heap usage info (available only if compiled with tcmalloc)");
+  ceph_assert(r == 0);
   return 0;
 }
 
@@ -3858,6 +3867,22 @@ will start to track new ops received afterwards.";
         goto out;
       }
     }
+  } else if (admin_command == "heap") {
+    if (!ceph_using_tcmalloc()) {
+      ss << "not using tcmalloc";
+      ret = -EOPNOTSUPP;
+      goto out;
+    }
+    string heapcmd;
+    cmd_getval(cmdmap, "heapcmd", heapcmd);
+    vector<string> heapcmd_vec;
+    get_str_vec(heapcmd, heapcmd_vec);
+    string value;
+    if (cmd_getval(cmdmap, "value", value)) {
+      heapcmd_vec.push_back(value);
+    }
+    ceph_heap_profiler_handle_command(heapcmd_vec, ss);
+    ret = 0;
   }
   dout(10) << "ret := " << ret << dendl;
   return true;
